### PR TITLE
pass in a FLUTTER_ENVIRONMENT env var when shelling out to the flutter cli

### PR DESCRIPTION
**Pull request by devoncarew**
- pass in a FLUTTER_ENVIRONMENT env var when shelling out to the flutter cli
- fix https://github.com/flutter/flutter-intellij/issues/840

@skybrian

### CODE REVIEW

File: src/io/flutter/sdk/FlutterSdkUtil.java [hunk 1]
```diff
@@ -35,13 +36,25 @@
 
 
 public class FlutterSdkUtil {
+  /** The environment variable to use to tell the flutter tool which app is driving it. */
+  public static final String FLUTTER_ENVIRONMENT = "FLUTTER_ENVIRONMENT";
```
Comment (skybrian) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: The name seems ambiguous - "environment" could mean anything. For the analysis server we use "client_id" and "client_version". Should we do the same here?

Comment (devoncarew) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: Perhaps `FLUTTER_CLIENT`? `FLUTTER_HOST`? This env var is meant to given the flutter cli tool some info about the environment it's running in: IntelliJ, Travis, ...

Comment (skybrian) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: FLUTTER_CLIENT sounds good. Do you want to pass in the version separately? Could be FLUTTER_CLIENT_ID and FLUTTER_CLIENT_VERSION.

(Or perhaps FLUTTER_CALLER?)

Comment (skybrian) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: This is much like the user-agent string in browsers. Hopefully we never have to use it for anything other than analytics.

Comment (devoncarew) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: Perhaps `FLUTTER_HOST_ENV` then? The var could represent IntelliJ, or it could represent Travis.

Comment (devoncarew) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: (will update to `FLUTTER_HOST_ENV`)

File: src/io/flutter/sdk/FlutterSdkUtil.java [hunk 2]
```diff
@@ -49,9 +49,9 @@ private FlutterSdkUtil() {
   /**
    * Return the environment variable value to use when shelling out to the Flutter command-line tool.
    */
-  public static String getFlutterEnvValue() {
+  public static String getFlutterHostEnvValue() {
     final String clientId = ApplicationNamesInfo.getInstance().getFullProductName().replaceAll(" ", "-");
```
Comment (skybrian) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: Maybe Flutter should handle replacing the space with a dash, so it's done consistently?

File: src/io/flutter/sdk/FlutterSdkUtil.java [hunk 3]
```diff
@@ -49,9 +49,9 @@ private FlutterSdkUtil() {
...
-    final String existingVar = System.getenv(FLUTTER_ENVIRONMENT);
+    final String existingVar = System.getenv(FLUTTER_HOST_ENV);
     return existingVar == null ? clientId : (existingVar + ":" + clientId);
```
Comment (skybrian) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: What's our allowed character set? How do we escape special characters? Maybe send JSON?

File: src/io/flutter/sdk/FlutterSdkUtil.java (hunk 2 above)
Comment (devoncarew) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: This is just normalizing what we're sending to the flutter cli - we'll mostly have control over what is send into the tool (we'll send it in from IntelliJ; similar modifications to the travis testing script and other callers will need to be made).

File: src/io/flutter/sdk/FlutterSdkUtil.java (hunk 3 above)
Comment (devoncarew) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: Not clear, but happy to put together a regex to whitelist some ranges of chars. From the IntelliJ does, `getProductName()` will return something like:

```
shortened name of the product if it contains two words (e.g. {@code "IDEA"} for IntelliJ IDEA, {@code "WebStorm"} for WebStorm)
```

Comment (skybrian) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: Okay, fair enough. Maybe this is over-engineering, but I was thinking we could either send a set of hashtags as JSON:
```
["travis","intellij-idea"]
```
Or a map as JSON:
```
{"travis":true,"product":"IntelliJ IDEA"}
```
And let the JSON encoder/parser worry about special characters. (Reasonably compact, and better than getting into an ad-hoc user-agent parsing situation later.)

Comment (devoncarew) [inline src/io/flutter/sdk/FlutterSdkUtil.java]: So, I think we're ok passing plain text here. We did do something very similar for invoking pub a few weeks ago. I don't think this graduate to a full user agent type thing, but just be used in analytics to get a better sense of how the flutter cli is used (exclude bots, only human cli usages and not tool driven, ...).

**Pull request merged**